lib/stdlib/src/maps.erl

```diff
@@ -90,35 +95,33 @@ intersect_combiner(_K, _V1, V2) ->
 intersect_with(Combiner, Map1, Map2) when is_map(Map1),
                                           is_map(Map2),
                                           is_function(Combiner, 3) ->
-    case map_size(Map1) < map_size(Map2) of
+    %% Use =< because we want to avoid reversing the combiner if we can
+    case map_size(Map1) =< map_size(Map2) of
         true ->
-            Iterator = maps:iterator(Map1),
-            intersect_with_1(maps:next(Iterator),
-                             Map1,
-                             Map2,
-                             Combiner);
+            intersect_with_sorted(Combiner, Map1, Map2);
         false ->
-            Iterator = maps:iterator(Map2),
-            intersect_with_1(maps:next(Iterator),
-                             Map2,
-                             Map1,
-                             fun(K, V1, V2) -> Combiner(K, V2, V1) end)
+            RCombiner = fun(K, V1, V2) -> Combiner(K, V2, V1) end,
+            intersect_with_sorted(RCombiner, Map2, Map1)
     end;
 intersect_with(Combiner, Map1, Map2) ->
     error(error_type_merge_intersect(Map1, Map2, Combiner),
           [Combiner, Map1, Map2]).
 
-intersect_with_1({K, V1, Iterator}, Map1, Map2, Combiner) ->
+intersect_with_sorted(Combiner, SmallMap, BigMap) ->
```

Comment:
The sorted word confused me a little bit as it made me think that the function would use sorting in some way before I realized that it just meant that the function expects that the parameters have some properties. Perhaps this can be explained in a comment or "_sorted" can be changed to "_smaller_map_first". I don't think this is a big issue so I'm ok with keeping it as it is if you think that is better.

Reply:
Oh, I like that, much clearer. I will change it.